### ISSUE TITLE
Mapped Type `[Lhs of Rhs]` Syntax for Array and Tuple Types

### ISSUE THREAD

**Issue by NonSpicyBurrito** [microsoft/TypeScript#54575]
# Suggestion

## 🔍 Search Terms

mapped type, of, array, tuple

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

A new `[Lhs of Rhs]` syntax in mapped type to map over values of array and tuple types:
```ts
type ResultType = {
    [V of SourceType]: DoSomethingWith<V>
}
```
Rhs does not need to be a type parameter.
Rhs must extend `readonly unknown[]` or a type error is raised.

## 📃 Motivating Example

```ts
type Foo = ['hello', 'world']

type Bar = {
    [V of Foo]: V[]
}
// Bar = ['hello'[], 'world'[]]
```

## 💻 Use Cases

Currently, to do the same as in motivating example, one might intuitively write:
```ts
type Bar = {
    [K in keyof Foo]: Foo[K][]
}
```
However this is incorrect and is a mistake that's potentially difficult to spot.

Instead one must either introduce a generic intermediate type:
```ts
type Transform<T> = {
    [K in keyof T]: T[K][]
}

type Bar = Transform<Foo>
```
Or use the `infer` workaround:
```ts
type Bar = Foo extends infer T ? {
    [K in keyof T]: T[K][]
} : never
```

Both workarounds are not immediately obvious to inexperienced developers, less ergonomic, and will silently break when `Foo` is no longer an array or tuple type without putting extra constraints.

**Comment by jcalz** [comment]
Presumably if #27995 were fixed we wouldn't need such a feature?  Is there a reason we'd want this feature *instead* of fixing #27995?

**Comment by NonSpicyBurrito** [comment]
> Presumably if #27995 were fixed we wouldn't need such a feature? Is there a reason we'd want this feature _instead_ of fixing #27995?

I would say both the fix and this feature can co-exist, as the second (ergonomics) and third (enforcing rhs must extend `readonly unknown[]`) points this feature provides are still valuable.
